sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/MapAggregate.scala

```diff
@@ -0,0 +1,310 @@
+  }
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    val defaultCheck = super.checkInputDataTypes()
+    if (defaultCheck.isFailure) {
+      defaultCheck
+    } else if (!numBinsExpression.foldable) {
+      TypeCheckFailure("The maximum number of bins provided must be a constant literal")
+    } else if (numBins < 2) {
+      val currentValue = if (numBinsExpression.eval() == null) null else numBins
+      TypeCheckFailure(
+        "The maximum number of bins provided must be a positive integer literal >= 2 " +
+          s"(current value = $currentValue)")
+    } else {
+      TypeCheckSuccess
+    }
+  }
+
+  override def update(buffer: MapDigest, input: InternalRow): Unit = {
+    if (!buffer.isInvalid) {
+      val evaluated = child.eval(input)
+      if (evaluated != null) buffer.update(child.dataType, evaluated, numBins)
+    }
+  }
+
+  override def merge(buffer: MapDigest, other: MapDigest): Unit = {
+    if (!buffer.isInvalid) {
+      if (other.isInvalid) {
+        buffer.isInvalid = true
+        buffer.clear()
+      } else {
+        buffer.merge(other, numBins)
+      }
+    }
+  }
+
+  override def eval(buffer: MapDigest): Any = {
+    if (buffer.isInvalid) {
+      // return empty map
+      ArrayBasedMapData(Map.empty)
+    } else {
+      // sort the result to make it more readable
+      val sorted = buffer match {
+        case stringDigest: StringMapDigest => TreeMap[UTF8String, Long](stringDigest.bins.toSeq: _*)
+        case numericDigest: NumericMapDigest => TreeMap[Double, Long](numericDigest.bins.toSeq: _*)
+      }
+      if (sorted.isEmpty) {
+        // don't have non-null values
+        null
+      } else {
+        ArrayBasedMapData(sorted.keys.toArray, sorted.values.toArray)
+      }
+    }
+  }
+
+  override def serialize(buffer: MapDigest): Array[Byte] = {
+    buffer.serialize()
+  }
+
+  override def deserialize(bytes: Array[Byte]): MapDigest = {
+    MapDigest.deserialize(child.dataType, bytes)
+  }
+
+  override def createAggregationBuffer(): MapDigest = {
+    child.dataType match {
+      case StringType => StringMapDigest()
+      case _ => NumericMapDigest()
+    }
+  }
+
+  override def withNewMutableAggBufferOffset(newMutableAggBufferOffset: Int): MapAggregate = {
+    copy(mutableAggBufferOffset = newMutableAggBufferOffset)
+  }
+
+  override def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): MapAggregate = {
+    copy(inputAggBufferOffset = newInputAggBufferOffset)
+  }
+
+  override def nullable: Boolean = true
+
+  override def dataType: DataType = {
+    child.dataType match {
+      case StringType => MapType(StringType, LongType)
+      case _ => MapType(DoubleType, LongType)
+    }
+  }
+
+  override def children: Seq[Expression] = Seq(child, numBinsExpression)
+
+  override def prettyName: String = "map_aggregate"
+}
+
+trait MapDigest {
+  // Mark this MapDigest invalid when the size of the hashmap (ndv of the column) exceeds numBins
```

Comment:
note - a basic thing: please use `/**  ...  */` for commenting member variables or methods, rather than `//`